docker/vep/vep.py

```diff
@@ -0,0 +1,227 @@
+            break
+
+        group.append(elt)
+
+    if group:
+        yield group
+
+
+def context():
+    return '''##fileformat=VCFv4.1
+#CHROM\tPOS\tID\tREF\tALT\tQUAL\tFILTER\tINFO\tFORMAT
+'''
+
+
+class Variant:
+    @staticmethod
+    def from_vcf_line(l):
+        fields = l.split('\t')
+        contig = fields[0]
+        pos = fields[1]
+        ref = fields[3]
+        alts = fields[4].split(',')
+        return Variant(contig, pos, ref, alts)
+
+    @staticmethod
+    def from_string(s):
+        fields = s.split(':')
+        contig = fields[0]
+        pos = fields[1]
+        ref = fields[2]
+        alts = fields[3].split(',')
+        return Variant(contig, pos, ref, alts)
+
+    def __init__(self, contig, pos, ref, alts):
+        self.contig = contig
+        self.position = pos
+        self.ref = ref
+        self.alts = alts
+
+    def to_vcf_line(self):
+        v = self.strip_star_allele()
+        return f'{v.contig}\t{v.position}\t.\t{v.ref}\t{",".join(v.alts)}\t.\t.\tGT\n'
+
+    def to_locus_alleles(self):
+        return ((self.contig, self.position), self.ref + self.alts)
+
+    def strip_star_allele(self):
+        return Variant(self.contig, self.position, self.ref, [a for a in self.alts if a != '*'])
+
+    def __str__(self):
+        return f'{self.contig}:{self.position}:{self.ref}:{",".join(self.alts)}'
+
+
+def consume_header(f) -> str:
+    header = ''
+    line = None
+    pos = 0
+    while line is None or line.startswith('#'):
+        pos = f.tell()
+        line = f.readline()
+        if line.startswith('#'):
+            header += line
+    f.seek(pos)
+    return header
+
+
+def get_csq_header(vep_cmd):
+    with sp.Popen(vep_cmd, env=os.environ, stdin=sp.PIPE, stdout=sp.PIPE, encoding='utf-8') as proc:
+        header = context()
+        v = Variant(1, 13372, 'G', ['C'])
+        data = f'{header}\n{v.to_vcf_line()}'
+
+        stdout, stderr = proc.communicate(data)
+        if stderr:
+            print(stderr)
+
+        for line in stdout.split('\n'):
+            line = line.rstrip()
+            for match in CONSEQUENCE_HEADER_REGEX.finditer(line):
+                return match.group(1)
+        print('WARNING: could not get VEP CSQ header')
+        return None
+
+
+def run_vep(vep_cmd, input_file, block_size, consequence, tolerate_parse_error, part_id, env):
+    results = []
+
+    with open(input_file, 'r') as inp:
+        header = consume_header(inp)
+        for block_id, block in enumerate(grouped_iterator(block_size, inp)):
+            n_processed = len(block)
+            start_time = time.time()
+
+            proc_id = f'{{"part_id":{part_id},"block_id":{block_id}}}'
+            variants = [Variant.from_vcf_line(l.rstrip()) for l in block]
+            non_star_to_orig_variants = {str(v.strip_star_allele()): str(v) for v in variants}
+
+            with sp.Popen(vep_cmd, env=env, stdin=sp.PIPE, stdout=sp.PIPE, stderr=sp.PIPE, encoding='utf-8') as proc:
```

Comment:
Hmm, curious, tell me more about this. Why can't we just run VEP on the VCF shard directly?

Reply:
I did the exact same thing as we do in the Scala implementation. Even if I had VEP call the variants from the VCF directly, we'd still need to parse the result in Python. We also keep a mapping of the original variant name with some preprocessing to match with the output from VEP.